### PR TITLE
[WIP] harvest-finance

### PR DESCRIPTION
**Pull request by froid1911**

### CODE REVIEW

File: harvest-finance/src/mapping.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    // simple transfer
+    let sender = getOrCreateAccount(event.params.from);
+    redeemFromMarket(
+      event,
+      sender,
+      market,
+      outputTokenAmount,
+      inputTokenAmounts,
+      rewardTokenAmounts,
+      outputTokenBalance,
```
Comment (gvladika) [inline harvest-finance/src/mapping.ts:138]: Careful here, we need balances of sender, not receiver :)

File: harvest-finance/src/mapping.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    [farmERC20], // FARM
+  )
+
+  market.save();
+
+  Vault.create(add.inputs._vault);
+}
+
+// update fAsset - mint/burn/transfer of users
+export function handleTransfer(event: Transfer): void {
```
Comment (gvladika) [inline harvest-finance/src/mapping.ts:50]: When mint or burn occurs market state changes as well - outputTokenBalance (fAsset amount) and inputTokenBalance(redeemable underlying asset) -> we need to have updateMarket executed

File: harvest-finance/src/mapping.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      outputTokenBalance,
+      inputTokenBalances,
+      rewardTokenBalances,
+      null
+    );
+  }
+}
+
+export function handleDoHardWork(call: DoHardWorkCall): void {
+  // create fake event
```
Comment (gvladika) [inline harvest-finance/src/mapping.ts:161]: You can simplify it this way

  let event = new ethereum.Event();
  event.block = call.block;
  event.transaction = call.transaction;

Comment (froid1911) [reply harvest-finance/src/mapping.ts:161]: not possible with apiVersion 0.0.6 :(

File: harvest-finance/src/mapping.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+  let market = Market.load(call.to.toHexString()) as Market;
+  let contract = VaultContract.bind(call.to);
+  let addressERC20InputResponse = contract.try_underlying();
+  if(addressERC20InputResponse.reverted) {
+    return;
+  }
+  let addressERC20Input = addressERC20InputResponse.value;
+
+  let pricePerShareResponse = contract.try_getPricePerFullShare();
```
[the remark this answers is not in the document]
Comment (froid1911) [reply harvest-finance/src/mapping.ts:184]: I think such an event does not exist?

File: harvest-finance/src/common.ts [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+    event,
+    sender,
+    market,
+    outputTokenAmount,
+    inputTokenAmounts,
+    rewardTokenAmounts,
+    outputTokenBalance,
+    inputTokenBalances,
+    rewardTokenBalances,
+    null
```
Comment (gvladika) [inline harvest-finance/src/common.ts:620]: same here, but receiver address

**Review by gvladika — COMMENTED** [review]
Nice work, looks good! So only reward tracking is left

**Pull request merged**